vault/vault.go

```diff
@@ -163,3 +176,181 @@ func (v EnvVault) read(path string) (*api.Secret, error) {
+//    "lease_duration" : 86400,
+//    "renewable" : true,
+//    "auth" : null,
+//    "wrap_info" : null
+// }
+
+// A VaultAWSCredsResponse represents a response from the Vault API itself
+// containing the AWS keys and tokens, etc.
+type VaultAWSCredsResponse struct {
+	RequestID string `json:"request_id"`
+	Data      struct {
+		SecretKey     string      `json:"secret_key"`
+		AccessKey     string      `json:"access_key"`
+		SecurityToken interface{} `json:"security_token"`
+	} `json:"data"`
+	LeaseID       string      `json:"lease_id"`
+	Warnings      interface{} `json:"warnings"`
+	LeaseDuration int         `json:"lease_duration"`
+	Renewable     bool        `json:"renewable"`
+	Auth          interface{} `json:"auth"`
+	WrapInfo      interface{} `json:"wrap_info"`
+}
+
+// A VaultAWSCredsLease is returned from GetAWSCredsLease
+type VaultAWSCredsLease struct {
+	Vars            []string
+	LeaseExpiryTime time.Time
+	LeaseID         string
+	Role            string
+}
+
+// GetAWSCredsLease calls the Vault API and asks for AWS creds for a particular role,
+// returning a string slice of vars of the form "VAR=value", and a
+func (v EnvVault) GetAWSCredsLease(role string) (*VaultAWSCredsLease, error) {
+	r := v.client.NewRequest("GET", "/v1/aws/creds/"+role)
+
+	resp, err := v.client.RawRequest(r)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("Failed to get AWS creds lease, got response code %d", resp.StatusCode)
+	}
+
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to read Vault response body: %w", err)
+	}
+
+	var creds VaultAWSCredsResponse
+	err = json.Unmarshal(data, &creds)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to unmarshal Vault response body: %w", err)
+	}
+
+	// Set up a padded expiry time to make sure we never run over
+	expiryTime := time.Now().UTC().Add(time.Duration(creds.LeaseDuration-360) * time.Second)
+
+	// Construct the AWS env vars
+	vars := []string{
+		fmt.Sprintf("AWS_ACCESS_KEY_ID=%s", creds.Data.AccessKey),
+		fmt.Sprintf("AWS_SECRET_ACCESS_KEY=%s", creds.Data.SecretKey),
+	}
+
+	return &VaultAWSCredsLease{
+		Vars:            vars,
+		LeaseExpiryTime: expiryTime,
+		LeaseID:         creds.LeaseID,
+		Role:            role,
+	}, nil
+}
+
+// RevokeAWSCreds calls Vault and revokes an existing lease on AWS credentials
+func (v EnvVault) RevokeAWSCredsLease(leaseID, role string) error {
+	log.Infof("Revoking AWS lease ID '%s' for role '%s' in Vault", leaseID, role)
+
+	r := v.client.NewRequest("PUT", "/v1/sys/leases/revoke")
+
+	bodyStruct := struct {
+		LeaseID string `json:"lease_id"`
+	}{
+		LeaseID: leaseID,
+	}
+
+	body, err := json.Marshal(bodyStruct)
+	if err != nil {
+		return fmt.Errorf("Unable to JSON encode lease revocation body: %w", err)
+	}
+
+	r.Body = bytes.NewBuffer(body)
+	resp, err := v.client.RawRequest(r)
+
+	if err != nil {
+		return err
```

Comment:
Might be worth doing `return fmt.Errorf("Unable to create vault request: ", err)`

Reply:
The errors come back with prefixes that make sense already. 👍